drv/cosmo-hf/src/hf.rs

```diff
@@ -0,0 +1,209 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use drv_hash_api::SHA256_SZ;
+use drv_hf_api::{
+    HfDevSelect, HfError, HfMuxState, HfPersistentData, HfProtectMode,
+};
+use idol_runtime::{
+    LeaseBufReader, LeaseBufWriter, Leased, LenLimit, NotificationHandler,
+    RequestError, R, W,
+};
+use ringbuf::ringbuf_entry_root as ringbuf_entry;
+use userlib::{task_slot, RecvMessage};
+
+use crate::{ServerImpl, Trace, PAGE_SIZE_BYTES, SECTOR_SIZE_BYTES};
+
+task_slot!(HASH, hash_driver);
+
+impl idl::InOrderHostFlashImpl for ServerImpl {
+    fn read_id(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<[u8; 20], RequestError<HfError>> {
+        Ok(self.flash_read_id())
+    }
+
+    fn capacity(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<usize, RequestError<HfError>> {
+        todo!()
+    }
+
+    /// Reads the STATUS_1 register from the SPI flash
+    fn read_status(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError<HfError>> {
+        Ok(self.read_flash_status())
+    }
+
+    fn bulk_erase(
+        &mut self,
+        _: &RecvMessage,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if !matches!(protect, HfProtectMode::AllowModificationsToSector0) {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        todo!()
+    }
+
+    fn page_program(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+        data: LenLimit<Leased<R, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr as usize / SECTOR_SIZE_BYTES == 0
```

Comment:
I am still confused by this line tbh. Is this `addr < SECTOR_SIZE_BYTES`? I'd suggest writing it that way (or using a range if you prefer) so that it doesn't look like a mistake.

Reply:
Yes, `addr < SECTOR_SIZE_BYTES` is more clear (and I'll add a comment as well)

Reply:
Yeah, I think that's what it is in the end: I think clippy would even complain about the range comparison because the lower end is guaranteed by the type.

Later there's a second line where we compare the division result to 1, so there we'd need to use `BYTES..BYTES*2` range inclusion check. But you're also absolutely right: A single less-than comparison is all we need here. (I think.)